### ISSUE TITLE
`clippy::unwrap-or-default` ignores MSRV

### ISSUE THREAD

**Issue by Gnurou** [rust-lang/rust-clippy#17379]
### Summary

This has been found in the [Linux kernel](https://lore.kernel.org/all/DJSET0S73YYD.2HCX79F9XAWF3@garyguo.net/), where setting `CONFIG_CC_OPTIMIZE_FOR_SIZE=y` (which passes `-Copt-level=s` to `rustc`) triggers the lint even though `Default` for `*const T` is stable since 1.88, and `msrv = "1.85"` is set in `.clippy.toml`:

```
warning: use of `unwrap_or` to construct default value
  --> ../rust/kernel/soc.rs:66:10
  |
66 |         .unwrap_or(core::ptr::null())
  |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `unwrap_or_default()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.96.0/index.html#unwrap_or_default
  = note: `-W clippy::unwrap-or-default` implied by `-W clippy::all`
  = help: to override `-W clippy::all` add `#[allow(clippy::unwrap_or_default)]`
```

The issue is not reproducible with `-Copt-level=2`: clippy decides "is `null()` equivalent to Default::default()?" by inspecting the optimized MIR of `<*const T as Default>::default` exported by the locally-built core, and at `-Copt-level=2` the call to `null()` is optimized out of that MIR, so the check fails.

So there seems to be two distinct bugs with `clippy::unwrap-or-default`:

- It is ignoring MSRV,
- Its outcome depends on the codegen flags of a dependency.

This looks related to issue #15183.

### Reproducer

Minimal reproducer proposed by Claude:

`test.rs`:
```rust
#![no_std]
pub fn f(o: Option<*const u8>) -> *const u8 {
    o.unwrap_or(core::ptr::null())
}
```

`cb.rs`:
```rust
#![feature(compiler_builtins)]
#![compiler_builtins]
#![no_std]
```

Then build and test with:
```bash
CORE=$(rustc --print sysroot)/lib/rustlib/src/rust/library/core/src/lib.rs
for lvl in s 2; do
  mkdir -p $lvl
  RUSTC_BOOTSTRAP=1 rustc --edition 2024 --crate-name core --crate-type rlib \
      -Copt-level=$lvl --emit=obj,metadata --out-dir $lvl $CORE
  RUSTC_BOOTSTRAP=1 rustc --crate-name compiler_builtins --crate-type rlib \
      --sysroot=/dev/null -L$lvl --emit=metadata --out-dir $lvl cb.rs
  clippy-driver --sysroot=/dev/null -L$lvl --crate-type rlib --emit=metadata --out-dir $lvl test.rs
done
```

On the `s` pass, the lint will trigger, but not on the `2` one.

### Version

```text
rustc 1.96.1 (31fca3adb 2026-06-26)
binary: rustc
commit-hash: 31fca3adb283cc9dfd56b49cdee9a96eb9c96ffd
commit-date: 2026-06-26
host: x86_64-unknown-linux-gnu
release: 1.96.1
LLVM version: 22.1.2
```

**Comment by hiSandog** [comment]
This is specific enough for a focused regression test around `clippy::unwrap-or-default` ignores MSRV. The test should set up the failing path, assert the intended behavior, and include the current error or bad output so the fix is not only verified manually. `clippy::unwrap-or-default` is a useful concrete anchor because it is named directly in the report.
